neo/Consensus/ConsensusService.cs

```diff
@@ -136,19 +179,63 @@ private void Log(string message, LogLevel level = LogLevel.Info)
 
         private void OnChangeViewReceived(ConsensusPayload payload, ChangeView message)
         {
-            if (message.NewViewNumber <= context.ExpectedView[payload.ValidatorIndex])
+            // We keep track of the payload hashes received in this block, and don't respond with recovery
+            // in response to the same payload that we already responded to previously.
+            // ChangeView messages include a Timestamp when the change view is sent, thus if a node restarts
+            // and issues a change view for the same view, it will have a different hash and will correctly respond
+            // again; however replay attacks of the ChangeView message from arbitrary nodes will not trigger an
+            // additonal recovery message response.
+            if (!knownHashes.Add(payload.Hash)) return;
+            if (message.NewViewNumber <= context.ViewNumber)
+            {
+                bool shouldSendRecovery = false;
+                // Limit recovery to sending from `f` nodes when the request is from a lower view number.
+                int allowedRecoveryNodeCount = context.F;
+                for (int i = 0; i < allowedRecoveryNodeCount; i++)
+                {
+                    var eligibleResponders = context.Validators.Length - 1;
+                    var chosenIndex = (payload.ValidatorIndex + i + message.NewViewNumber) % eligibleResponders;
+                    if (chosenIndex >= payload.ValidatorIndex) chosenIndex++;
```

Comment:
I did not get this point here. Why to jump one here? The next iteration of the loop would play this role.

Reply:
Because the `ValidatorIndex` is the requester’s index. The requested does not respond to his own `ChangeView` message with a recovery.

Reply:
If instead it moved to the next iteration of the loop that would mean there would be one less potential responders. It is preferable to increment the `chosenIndex` to allow the correct number of nodes to respond with recovery.